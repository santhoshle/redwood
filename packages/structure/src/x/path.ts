import { existsSync } from 'fs'
import { basename, normalize, sep } from 'path'

export function directoryNameResolver(dirName: string): string | undefined {
  dirName = normalize(dirName)
  const parts = dirName.split(sep)
  const pp = parts[parts.length - 1]
  parts.push(pp)
  const extensions = ['.js', '.jsx', '.ts', '.tsx']
  const pathNoExt = parts.join(sep)
  for (const ext of extensions) {
    const path = pathNoExt + ext
    if (existsSync(path)) {
      return path
    }
  }
}

export function followsDirNameConvention(filePath: string): boolean {
  filePath = normalize(filePath)
  const ending = basenameNoExt(filePath) + sep + basename(filePath)
  return filePath.endsWith(ending)
}

export function basenameNoExt(path: string): string {
  path = normalize(path)
  const parts = basename(path).split('.')
  if (parts.length > 1) {
    parts.pop()
  }
  return parts.join('.')
}

export function isLayoutFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Layout')
}

export function isCellFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Cell')
}
