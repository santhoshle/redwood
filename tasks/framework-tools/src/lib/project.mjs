/* eslint-env node */

import fs from 'fs'
import path from 'path'

import execa from 'execa'
import ora from 'ora'
import rimraf from 'rimraf'
import terminalLink from 'terminal-link'

import {
  frameworkDependencies,
  frameworkPkgJsonFiles,
  frameworkPackagesFiles,
  frameworkPackagesBins,
  REDWOOD_PACKAGES_PATH,
} from './framework.mjs'

/**
 * Sets binaries as executable and creates symlinks to `node_modules/.bin` if they do not exist.
 */
export function fixProjectBinaries(projectPath) {
  const bins = frameworkPackagesBins()
  for (let [binName, binPath] of Object.entries(bins)) {
    // if the binPath doesn't exist, create it.
    const binSymlink = path.join(projectPath, 'node_modules/.bin', binName)
    binPath = path.join(projectPath, 'node_modules', binPath)
    if (!fs.existsSync(binSymlink)) {
      fs.mkdirSync(path.dirname(binSymlink), {
        recursive: true,
      })
      fs.symlinkSync(binPath, binSymlink)
    }
    console.log('chmod +x', terminalLink(binName, binPath))
    fs.chmodSync(binPath, '755')
  }
}

/**
 * Append all the `@redwoodjs` dependencies to the root `package.json` in a Redwood Project.
 */
export function addDependenciesToPackageJson(
  packageJsonPath,
  dependencenies = frameworkDependencies()
) {
  if (!fs.existsSync(packageJsonPath)) {
    console.log(
      `Error: The package.json path: ${packageJsonPath} does not exist.`
    )
    process.exit(1)
  }

  const packageJsonLink = terminalLink(
    'package.json',
    'file://' + packageJsonPath
  )

  const numOfDeps = Object.keys(dependencenies).length

  const spinner = ora(
    `Adding ${numOfDeps} framework dependencies to ${packageJsonLink}...`
  ).start()

  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'))
  packageJson.dependencies = {
    ...(packageJson.dependencies || {}),
    ...dependencenies,
  }
  fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, undefined, 2))
  spinner.succeed(
    `Added ${numOfDeps} framework dependencies to ${packageJsonLink}`
  )
}

export function installProjectPackages(projectPath) {
  const spinner = ora("Running 'yarn install'...")
  spinner.start()
  try {
    execa.sync('yarn install', {
      cwd: projectPath,
      shell: true,
    })
    spinner.succeed("Ran 'yarn install'")
  } catch (e) {
    spinner.warn(
      `Error running 'yarn install', check ${terminalLink(
        'yarn-error.log',
        'file://' + path.join(projectPath, 'yarn-error.log')
      )} for more information.`
    )
    console.log('-'.repeat(80))
  }
}

export function copyFrameworkFilesToProject(
  projectPath,
  packages = frameworkPkgJsonFiles()
) {
  // Loop over every package, delete all existing files, copy over the new files,
  // and fix binaries.
  packages = frameworkPackagesFiles(packages)
  for (const [packageName, files] of Object.entries(packages)) {
    const packageDstPath = path.join(projectPath, 'node_modules', packageName)
    console.log(
      terminalLink(packageName, 'file://' + packageDstPath),
      files.length,
      'files'
    )
    rimraf.sync(packageDstPath)

    for (const file of files) {
      const src = path.join(
        REDWOOD_PACKAGES_PATH,
        packageName.replace('@redwoodjs', ''),
        file
      )
      const dst = path.join(packageDstPath, file)
      fs.mkdirSync(path.dirname(dst), { recursive: true })
      fs.copyFileSync(src, dst)
    }
  }

  console.log()
  fixProjectBinaries(projectPath)
}
