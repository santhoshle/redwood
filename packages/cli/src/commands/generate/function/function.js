import path from 'path'

import camelcase from 'camelcase'
import Listr from 'listr'
import terminalLink from 'terminal-link'

import { getPaths, transformTSToJS, writeFilesTask } from 'src/lib'
import c from 'src/lib/colors'

import { yargsDefaults } from '../../generate'
import { templateForComponentFile } from '../helpers'

export const files = ({
  name,
  typescript: generateTypescript = false,
  ...rest
}) => {
  // Taken from ../component; should be updated to take from the project's configuration
  const extension = generateTypescript ? '.ts' : '.js'

  const functionName = camelcase(name)
  const file = templateForComponentFile({
    name: functionName,
    componentName: functionName,
    extension,
    apiPathSection: 'functions',
    generator: 'function',
    templatePath: 'function.ts.template',
    templateVars: { ...rest },
    outputPath: path.join(
      getPaths().api.functions,
      `${functionName}${extension}`
    ),
  })

  const template = generateTypescript
    ? file[1]
    : transformTSToJS(file[0], file[1])

  return { [file[0]]: template }
}

export const command = 'function <name>'
export const description = 'Generate a Function'

// This could be built using createYargsForComponentGeneration;
// however, functions wouldn't have a `stories` option. createYargs...
// should be reversed to provide `yargsDefaults` as the default configuration
// and accept a configuration such as its CURRENT default to append onto a command.
export const builder = (yargs) => {
  yargs
    .positional('name', {
      description: 'Name of the Function',
      type: 'string',
    })
    .epilogue(
      `Also see the ${terminalLink(
        'Redwood CLI Reference',
        'https://redwoodjs.com/reference/command-line-interface#generate-function'
      )}`
    )

  // Add default options, includes '--typescript', '--javascript', '--force', ...
  Object.entries(yargsDefaults).forEach(([option, config]) => {
    yargs.option(option, config)
  })
}

// This could be built using createYargsForComponentGeneration;
// however, we need to add a message after generating the function files
export const handler = async ({ name, ...rest }) => {
  const tasks = new Listr(
    [
      {
        title: `Generating function files...`,
        task: async () => {
          const f = await files({ name, ...rest })
          return writeFilesTask(f, { overwriteExisting: rest.force })
        },
      },
    ],
    { collapse: false, exitOnError: true }
  )

  try {
    await tasks.run()

    console.info('')
    console.info(c.warning('âš  Important:'))
    console.info('')

    console.info(
      c.bold(
        'When deployed, a custom serverless function is an open API endpoint and ' +
          'is your responsibility to secure appropriately.'
      )
    )

    console.info('')
    console.info(
      `Please consult the ${terminalLink(
        'Serverless Function Considerations',
        'https://redwoodjs.com/docs/serverless-functions#security-considerations'
      )} in the RedwoodJS documentation for more information.`
    )
    console.info('')
  } catch (e) {
    console.error(c.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
