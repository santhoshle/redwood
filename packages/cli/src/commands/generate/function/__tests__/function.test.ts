global.__dirname = __dirname
import path from 'path'

// Load shared mocks
import 'src/lib/test'

import * as functionGenerator from '../function'

// Should be refactored as it's repeated
type WordFilesType = { [key: string]: string }

let singleWordDefaultFiles: WordFilesType
let multiWordDefaultFiles: WordFilesType
let javascriptFiles: WordFilesType
let typescriptFiles: WordFilesType

beforeAll(() => {
  singleWordDefaultFiles = functionGenerator.files({
    name: 'foo',
  })
  multiWordDefaultFiles = functionGenerator.files({
    name: 'send-mail',
  })
  javascriptFiles = functionGenerator.files({
    name: 'javascript-function',
  })
  typescriptFiles = functionGenerator.files({
    name: 'typescript-function',
    typescript: true,
  })
})

test('returns exactly 1 file', () => {
  expect(Object.keys(singleWordDefaultFiles).length).toEqual(1)
})

test('creates a single word function file', () => {
  expect(
    singleWordDefaultFiles[
      path.normalize('/path/to/project/api/src/functions/foo.js')
    ]
  ).toMatchSnapshot()
})

test('creates a multi word function file', () => {
  expect(
    multiWordDefaultFiles[
      path.normalize('/path/to/project/api/src/functions/sendMail.js')
    ]
  ).toMatchSnapshot()
})

test('creates a .js file if --javascript=true', () => {
  expect(
    javascriptFiles[
      path.normalize('/path/to/project/api/src/functions/javascriptFunction.js')
    ]
  ).toMatchSnapshot()
  // ^ JS-function-args should be stripped of their types and consequently the unused 'aws-lamda' import removed.
  // https://babeljs.io/docs/en/babel-plugin-transform-typescript
})

test('creates a .ts file if --typescript=true', () => {
  expect(
    typescriptFiles[
      path.normalize('/path/to/project/api/src/functions/typescriptFunction.ts')
    ]
  ).toMatchSnapshot()
  // ^ TS-functions, on the other hand, retain the 'aws-lamda' import and type-declartions.
})
