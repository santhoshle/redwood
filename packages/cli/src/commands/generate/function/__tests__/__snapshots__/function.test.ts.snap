// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a .js file if --javascript=true 1`] = `
"import { logger } from 'src/lib/logger'

/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 *
 * Important: When deployed, a custom serverless function is an open API endpoint and
 * is your responsibility to secure appropriately.
 *
 * @see {@link https://redwoodjs.com/docs/serverless-functions#security-considerations|Serverless Function Considerations}
 * in the RedwoodJS documentation for more information.
 *
 * @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent
 * @typedef { import('aws-lambda').Context } Context
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event, context) => {
  logger.info('Invoked javascriptFunction function')

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },

    body: JSON.stringify({
      data: 'javascriptFunction function',
    }),
  }
}
"
`;

exports[`creates a .ts file if --typescript=true 1`] = `
"import type { APIGatewayEvent, Context } from 'aws-lambda'
import { logger } from 'src/lib/logger'

/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 *
 * Important: When deployed, a custom serverless function is an open API endpoint and
 * is your responsibility to secure appropriately.
 *
 * @see {@link https://redwoodjs.com/docs/serverless-functions#security-considerations|Serverless Function Considerations}
 * in the RedwoodJS documentation for more information.
 *
 * @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent
 * @typedef { import('aws-lambda').Context } Context
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event: APIGatewayEvent, context: Context) => {
  logger.info('Invoked typescriptFunction function')

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      data: 'typescriptFunction function',
    }),
  }
}
"
`;

exports[`creates a multi word function file 1`] = `
"import { logger } from 'src/lib/logger'

/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 *
 * Important: When deployed, a custom serverless function is an open API endpoint and
 * is your responsibility to secure appropriately.
 *
 * @see {@link https://redwoodjs.com/docs/serverless-functions#security-considerations|Serverless Function Considerations}
 * in the RedwoodJS documentation for more information.
 *
 * @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent
 * @typedef { import('aws-lambda').Context } Context
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event, context) => {
  logger.info('Invoked sendMail function')

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },

    body: JSON.stringify({
      data: 'sendMail function',
    }),
  }
}
"
`;

exports[`creates a single word function file 1`] = `
"import { logger } from 'src/lib/logger'

/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 *
 * Important: When deployed, a custom serverless function is an open API endpoint and
 * is your responsibility to secure appropriately.
 *
 * @see {@link https://redwoodjs.com/docs/serverless-functions#security-considerations|Serverless Function Considerations}
 * in the RedwoodJS documentation for more information.
 *
 * @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent
 * @typedef { import('aws-lambda').Context } Context
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event, context) => {
  logger.info('Invoked foo function')

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },

    body: JSON.stringify({
      data: 'foo function',
    }),
  }
}
"
`;
